### DISCUSSION TITLE
Close inactive connections after a delay

### DISCUSSION THREAD

**Discussion by nmoreaud** [sqlalchemy/sqlalchemy#11943] [Usage Questions]
Hello
I wonder if I could easily close inactive SQL connections after a delay.
There is the `pool_recycle` option, but it is only used during checkout, and automatically replaces an old connection.
Is there a simple way to close unused connections?

**Comment by zzzeek** [comment]
to close them from the client side after a delay is not that simple since you would need to run a background thread to do this.  it would have to check connections out of the pool, determine if they are past a certain age, then hard close them using `connection.detach()`.

to close from the server side may or may not be simple depending on if your database has a server side usage timeout, however, the client wont know the connection has been severed until next checkout.  using pre-ping would allow this usage to be seamless.

**Reply by nmoreaud** [reply]
Ok thanks, I'll see if we do it.

How can I check every connection out of the pool without opening new connections or hang?
Should I subscribe to `sqlalchemy.events.PoolEvents.connect` to maintain a list of existing connections?

**Reply by zzzeek** [reply]
yah that might actually be a good idea, using the connect event as a marker for age.   try that
